torch_glow/src/CachingGraphRunner.cpp

```diff
@@ -20,61 +20,128 @@
+                    const torch::jit::Stack &stack) {
+  const std::shared_ptr<torch::jit::Graph> graph = node->g(at::attr::Subgraph);
+  const auto &graphInputs = graph->inputs();
+  const auto inputs = torch::jit::last(stack, graphInputs.size());
+  torch::jit::CompleteArgumentSpec spec(/*with_grad*/ false, inputs);
+
+  return torch::hash_combine(reinterpret_cast<size_t>(node), spec.hashCode());
+}
+
+} // namespace
 
+llvm::Expected<CachingGraphRunner::PerGlowGraphInfo *>
+CachingGraphRunner::loadGraphImpl(const torch::jit::Node *node,
+                                  torch::jit::Stack &stack) {
   const std::shared_ptr<torch::jit::Graph> graph = node->g(at::attr::Subgraph);
+  RETURN_ERR_IF_NOT(graph, "A subgraph couldn't be found for this node");
+
   const auto &graphInputs = graph->inputs();
-  const auto numInputs = graphInputs.size();
-  auto inputs = torch::jit::last(stack, numInputs);
-  const char *const functionName = "PyTorchFunction";
+  auto inputs = torch::jit::last(stack, graphInputs.size());
+
+  size_t hash = getGraphHash(node, stack);
+
+  // If we already have a Glow function compiled for this graph with and the
+  // given inputs then use that.
+  auto it = perGlowGraphInfoMap.find(hash);
+  if (it != perGlowGraphInfoMap.end()) {
+    return it->second.get();
+  }
+
+  auto info = llvm::make_unique<PerGlowGraphInfo>();
+  info->functionName = strFormat("PTFunction%lu", hash);
+  info->hash = hash;
+  info->node = node;
 
-  glow::Function *f = nullptr;
-  executionEngine_.setBackendName(executionEngine_.getBackendName());
-  auto &mod = executionEngine_.getModule();
-  f = mod.createFunction(functionName);
-  std::vector<glow::Placeholder *> inputPlaceholders;
-  std::vector<glow::Placeholder *> outputPlaceholders;
+  std::unique_ptr<Module> module = llvm::make_unique<Module>();
+  Function *f = module->createFunction(info->functionName);
 
   RETURN_IF_ERR(PyTorchModelLoader::loadJITGraph(
-      *f, *graph, inputs, inputPlaceholders, outputPlaceholders,
+      *f, *graph, inputs, info->inputPlaceholders, info->outputPlaceholders,
       getPyTorchLoaderSettings()));
 
   glow::CompilationContext cctx;
-  executionEngine_.compile(cctx);
 
-  glow::PlaceholderBindings bindings;
-  for (size_t i = 0; i < inputs.size(); ++i) {
-    glow::Placeholder *ph = inputPlaceholders[i];
+  RETURN_IF_ERR(hostManager_->addNetwork(std::move(module), cctx));
+
+  perGlowGraphInfoMap.insert({hash, std::move(info)});
+
+  return perGlowGraphInfoMap[hash].get();
+}
+
+llvm::Error CachingGraphRunner::runGraphImpl(const PerGlowGraphInfo &info,
+                                             torch::jit::Stack &stack) {
+  DCHECK_EQ(getGraphHash(info.node, stack), info.hash)
+      << "Tried to run a graph with incompatible input shapes.";
+
+  size_t numInputs = info.inputPlaceholders.size();
+
+  auto inputs = torch::jit::last(stack, numInputs);
+
+  std::unique_ptr<ExecutionContext> ctx = llvm::make_unique<ExecutionContext>();
```

Comment:
It will create a bindings for every run, is it possible to cache bindings as well?

Reply:
I'm not sure it'll be too much of a win, haven't noticed this being an issue in onnxifi

Reply:
Yeah, input/output tensor usually have reasonable sizes.